packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx

```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
-          {descriptor instanceof Descriptor ? (
-            <PresentationInstanceFilterBuilder {...restProps} descriptor={descriptor} onInstanceFilterChanged={onInstanceFilterChanged} />
-          ) : (
-            <DelayLoadedPresentationInstanceFilterBuilder {...restProps} onInstanceFilterChanged={onInstanceFilterChanged} descriptorGetter={descriptor} />
-          )}
-        </Dialog.Content>
-        <div className="presentation-instance-filter-dialog-bottom-container">
-          <div>{filterResultCountRenderer && filterResultCountRenderer(filter)}</div>
-          <Dialog.ButtonBar className="presentation-instance-filter-button-bar">
-            <Button className="presentation-instance-filter-dialog-apply-button" styleType="high-visibility" onClick={applyButtonHandle} disabled={!filter}>
-              {translate("instance-filter-builder.apply")}
-            </Button>
-            <Button className="presentation-instance-filter-dialog-close-button" onClick={onClose}>
-              {translate("instance-filter-builder.cancel")}
-            </Button>
-          </Dialog.ButtonBar>
-        </div>
+        {descriptor instanceof Descriptor ? (
+          <PresentationInstanceFilterDialogContent {...restProps} descriptor={descriptor} onClose={onClose} />
+        ) : (
+          <DelayedCenteredProgressRadial />
+        )}
       </Dialog.Main>
     </Dialog>
   );
 }
 
-interface DelayLoadedPresentationInstanceFilterBuilderProps extends Omit<PresentationInstanceFilterBuilderProps, "descriptor"> {
-  descriptorGetter: () => Promise<Descriptor>;
-}
-
-function DelayLoadedPresentationInstanceFilterBuilder(props: DelayLoadedPresentationInstanceFilterBuilderProps) {
-  const { descriptorGetter, ...restProps } = props;
-  const descriptor = useDelayLoadedDescriptor(descriptorGetter);
-  if (!descriptor) {
-    return <DelayedCenteredProgressRadial />;
-  }
-
-  return <PresentationInstanceFilterBuilder {...restProps} descriptor={descriptor} />;
-}
-
-function useDelayLoadedDescriptor(descriptorGetter: () => Promise<Descriptor>) {
-  const [descriptor, setDescriptor] = useState<Descriptor>();
+function useDelayLoadedDescriptor(descr: Descriptor | (() => Promise<Descriptor>)) {
+  const [descriptor, setDescriptor] = useState<Descriptor | (() => Promise<Descriptor>)>(() => descr);
+  const disposed = useRef<boolean>(false);
 
   useEffect(() => {
-    let disposed = false;
     void (async () => {
-      const newDescriptor = await descriptorGetter();
-      // istanbul ignore else
-      if (!disposed) {
-        setDescriptor(newDescriptor);
+      if (descriptor && !(descriptor instanceof Descriptor)) {
+        const newDescriptor = await descriptor();
+        // istanbul ignore else
+        if (!disposed.current) {
+          setDescriptor(newDescriptor);
+        }
       }
     })();
     return () => {
-      disposed = true;
+      disposed.current = true;
     };
-  }, [descriptorGetter]);
+  }, [descr, descriptor]);
```

Comment:
You shouldn't use `descriptor` state variable inside this `useEffect`

Reply:
`npm run lint` shows an error then
![image](https://github.com/iTwin/presentation/assets/124659623/84d262c8-da70-4103-a637-8033dbbbf829)
added a comment to disable this rule